lib/posix/pthread_key.c

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <kernel.h>
+#include <posix/pthread.h>
+#include <posix/pthread_key.h>
+
+struct k_sem pthread_key_sem;
+
+K_SEM_DEFINE(pthread_key_sem, 1, 1);
+
+/**
+ * @brief Create a key for thread-specific data
+ *
+ * See IEEE 1003.1
+ */
+int pthread_key_create(pthread_key_t *key,
+		void (*destructor)(void *))
+{
+	pthread_key_obj *new_key;
+
+	*key = NULL;
+
+	new_key = k_malloc(sizeof(pthread_key_obj));
+
+	if (new_key == NULL) {
+		return ENOMEM;
+	}
+
+	sys_dlist_init(&(new_key->key_data_l));
+
+	new_key->destructor = destructor;
+	*key = (void *)new_key;
+
+	return 0;
+}
+
+/**
+ * @brief Delete a key for thread-specific data
+ *
+ * See IEEE 1003.1
+ */
+int pthread_key_delete(pthread_key_t key)
+{
+	pthread_key_obj *key_obj = (pthread_key_obj *)key;
+	pthread_key_data *key_data;
+	sys_dnode_t *node_l, *next_node_l;
+
+	k_sem_take(&pthread_key_sem, K_FOREVER);
```

Comment:
Conversely, why use a k_sem here instead of a pthread mutex, which you're using above for the same kind of lock?

Reply:
When I took a look at the specific implementation of POSIX functions like message queues, pthread_rwlock, etc, they directly made use of k_sem objects, and so I followed the same for pthread_key.c